### PR TITLE
Broke out OwnedBuffer of Text for reuse in other situations.

### PR DESCRIPTION
**Pull request by simonmika**

### CODE REVIEW

File: source/base/OwnedBuffer.ooc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+OwnedBuffer: cover {
+	_pointer: UInt8*
+	_size: Int
+	_owner: Owner
+	pointer ::= this _pointer
+	size ::= this _size
+	init: func@ {
+		this _pointer = null
+		this _size = 0
+		this _owner = Owner Unknown
```
Comment (davidhesselbom) [inline source/base/OwnedBuffer.ooc]: Why not make these the default values?

File: source/base/OwnedBuffer.ooc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		}
+		result
+	}
+	free: func@ ~withCriteria (criteria: Owner) -> Bool {
+		this _owner == criteria && this free()
+	}
+	copy: func -> This {
+		this _copyData()
+		this _owner = Owner Callee
+		this
```
Comment (davidhesselbom) [inline source/base/OwnedBuffer.ooc]: This looks really scary, but that's because I'm thinking of it as a class. Interesting stuff!

**Comment by davidhesselbom** [comment]
I don't think you meant to add `dataFile.txt` and `example.svg`. Amend.

**Comment by davidpiuva** [comment]
I always get those files when getting the latest Kean.

**Comment by simonmika** [comment]
I did indeed not. Can somebody, please fix the tests so that they always clean up their garbage.

File: source/base/OwnedBuffer.ooc (hunk 2 above)
Comment (simonmika) [inline source/base/OwnedBuffer.ooc]: We have been adding comments to help people out.

**Comment by davidhesselbom** [comment]
Gimme a sec...

File: source/base/OwnedBuffer.ooc (hunk 1 above)
Comment (simonmika) [inline source/base/OwnedBuffer.ooc]: Why initialize and then overwrite immediately? Covers are not zero the way objects are. But we should be calling another constructor.

**Comment by davidhesselbom** [comment]
Should be fixed in #449.

**Comment by simonmika** [comment]
It turned out that `Text`did not follow the rules for freeing.

File: source/base/TextBuffer.ooc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	operator []= (index: Int, value: Char) {
+		this raw[index] = value
+	}
+	operator []= (range: Range, data: This) {
+		data copyTo(this[range])
+	}
+	operator == (other: This) -> Bool {
+		this _backend == other _backend
+	}
+	empty: static This { get { This new() } }
```
Comment (davidhesselbom) [inline source/base/TextBuffer.ooc]: There are many places in our codebase where `empty` is a property that returns a `Bool`. Should those be changed into `isEmpty`? Either way, `empty` is a an adjective and a verb. Should it be obvious that
- `This empty` returns an "empty" instance of something,
- `this empty` returns a bool stating whether an instance is "empty", and
- `this empty()` "empties" an instance?

Comment (simonmika) [inline source/base/TextBuffer.ooc]: Is this relevant to this PR? Lets discuss the `empty` / `isEmpty` property outside the scope of this PR.

Comment (davidhesselbom) [inline source/base/TextBuffer.ooc]: It is relevant if we conclude that name of this property should change, but alright.

**Comment by davidhesselbom** [comment]
See #475.

File: source/base/OwnedBuffer.ooc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import math
+
+OwnedBuffer: cover {
+	_pointer: UInt8*
+	_size: Int
+	_owner: Owner
+	pointer ::= this _pointer
+	size ::= this _size
+	owner ::= this _owner
+	isOwned ::= this _owner != Owner Unknown && this _owner != Owner Static && this _owner != Owner Stack && this _pointer != null
```
Comment (davidhesselbom) [inline source/base/OwnedBuffer.ooc]: Would it make sense to put the part of this logic that compares `this _owner` to various `Owner` variants in its own function?

**Comment by davidhesselbom** [comment]
Aside from the single-letter variable names here and there (`t`, `n`, `s`, `c`), I guess this is ready. I can't say I understand everything that goes on in here, but it looks very cool...

File: source/base/OwnedBuffer.ooc (hunk 4 above)
Comment (simonmika) [inline source/base/OwnedBuffer.ooc]: `owner` is exposed as a property as well.

**Comment by simonmika** [comment]
@sebastianbaginski can you have a look at this code and ensure I got it right?

File: source/base/OwnedBuffer.ooc [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+			this _owner = Owner Caller
+		} else
+			result = this
+		result
+	}
+	copyTo: func (destination: This) -> Int {
+		result := 0
+		if (this _pointer != null && this _size != 0 && destination _pointer != null && destination _size != 0) {
+			result = Int minimum(this _size, destination _size)
+			memcpy(destination _pointer, this _pointer, result)
```
Comment (sebastianbaginski) [inline source/base/OwnedBuffer.ooc]: What if `destination` is a string literal ? I think the memory can't be written to in that case.

File: source/base/Text.ooc [hunk 6]
```diff
@@ -302,5 +312,4 @@ Text: cover {
 
 operator == (left: String, right: Text) -> Bool { Text new(left) == right }
 operator != (left: String, right: Text) -> Bool { !(left == right) }
-operator != (left: Text, right: String) -> Bool { !(left == right) }
-operator != (left: Text, right: Text) -> Bool { !(left == right) }
+operator []= (left: TextBuffer, range: Range, right: Text) { right _buffer copyTo(left slice(range min, range max - range min)) }
```
Comment (sebastianbaginski) [inline source/base/Text.ooc]: `range count`?

File: source/base/TextBuilder.ooc [hunk 7]
```diff
@@ -46,71 +48,73 @@ TextBuilder: class {
 		This new(this)
 	}
 	append: func ~char (c: Char) {
-		this append(Text new(c&, 1, Owner Stack) give())
+		this append(Text new(c&, 1, Owner Stack) copy())
```
Comment (sebastianbaginski) [inline source/base/TextBuilder.ooc]: `Text new(c)`?

File: source/base/TextBuilder.ooc [hunk 8]
```diff
@@ -46,71 +48,73 @@ TextBuilder: class {
...
+		this _data add(value claim())
 	}
 	append: func (other: This) {
 		for (i in 0 .. other count)
-			this append(other[i] copy())
+			this append(other _data[i] copy())
 	}
 	prepend: func ~char (c: Char) {
-		this prepend(Text new(c&, 1, Owner Stack) give())
+		this prepend(Text new(c&, 1, Owner Stack) copy())
```
Comment (sebastianbaginski) [inline source/base/TextBuilder.ooc]: `Text new(c)`?
